File: backend/src/main/java/com/neighborhood/eventmanagement/service/EventService.java
package com.neighborhood.eventmanagement.service;

import com.neighborhood.eventmanagement.dto.EventRequest;
import com.neighborhood.eventmanagement.dto.EventResponse;
import com.neighborhood.eventmanagement.entity.EventCategory;
import com.neighborhood.eventmanagement.entity.EventStatus;
import com.neighborhood.eventmanagement.entity.Zone;
import org.springframework.data.domain.Page;

import java.time.LocalDateTime;
import java.util.List;

public interface EventService {

    // =====================================================
    // CRUD
    // =====================================================

    EventResponse createEvent(EventRequest request);

    EventResponse updateEvent(Long id, EventRequest request);

    void deleteEvent(Long id);

    EventResponse getEventById(Long id);
Page<EventResponse> getEvents(
        int page,
        int size,
        String sortBy,
        String sortDirection,
        EventCategory category,
        EventStatus status,
        Long zoneId
);

    // =====================================================
    // PAGINATION + FILTERING (Issue #4)
    // =====================================================


    // =====================================================
    // SEARCH
    // =====================================================

    Page<EventResponse> searchEvents(
            String keyword,
            int page,
            int size,
            String sortBy,
            String sortDirection
    );

    // =====================================================
    // SRS APIs
    // =====================================================

    List<EventResponse> getMyEvents(Long organizerId);

    List<EventResponse> getUpcomingEvents();

    List<EventResponse> getEventsByCategory(EventCategory category);

    List<EventResponse> getCalendarEvents(
            Zone zone,
            LocalDateTime start,
            LocalDateTime end
    );

    // =====================================================
    // VALIDATION
    // =====================================================

    void validateEvent(EventRequest request);

}

File: frontend/src/services/eventService.js
import axios from "axios";

const API_BASE_URL = "http://localhost:8080/api/events";

// Strip keys with empty-string or undefined values so backend doesn't
// receive spurious filter params like category=&status=
function clean(params) {
    return Object.fromEntries(
        Object.entries(params).filter(([, v]) => v !== "" && v !== undefined && v !== null)
    );
}

const eventService = {

    getEvents(params) {
        return axios.get(API_BASE_URL, {
            params: clean({
                page: params.page,
                size: params.size,
                sortBy: params.sortBy,
                sortDirection: params.sortDirection,
                category: params.category,
                status: params.status,
                zoneId: params.zoneId
            })
        });
    },

    searchEvents(params) {
        return axios.get(`${API_BASE_URL}/search`, {
            params: clean({
                keyword: params.keyword,
                page: params.page,
                size: params.size,
                sortBy: params.sortBy,
                sortDirection: params.sortDirection
            })
        });
    },

    getEventById(id) {
        return axios.get(`${API_BASE_URL}/${id}`);
    },

    createEvent(event) {
        return axios.post(API_BASE_URL, event);
    },

    updateEvent(id, event) {
        return axios.put(`${API_BASE_URL}/${id}`, event);
    },

    deleteEvent(id) {
        return axios.delete(`${API_BASE_URL}/${id}`);
    }
};

export default eventService;
